subsys/net/ip/tcp2.c

```diff
@@ -267,11 +272,164 @@ int net_tcp_unref(struct net_context *context)
+
+	th->th_off = 5;
+	th->th_flags = flags;
+	th->th_win = htons(conn->recv_win);
+	th->th_seq = htonl(conn->seq);
+
+	if (ACK & flags) {
+		th->th_ack = htonl(conn->ack);
+	}
+
+	return net_pkt_set_data(pkt, &tcp_access);
+}
+
+static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
+{
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		return net_context_create_ipv4_new(conn->context, pkt,
+						&conn->src.sin.sin_addr,
+						&conn->dst.sin.sin_addr);
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
+		return net_context_create_ipv6_new(conn->context, pkt,
+						&conn->src.sin6.sin6_addr,
+						&conn->dst.sin6.sin6_addr);
+	}
+
+	return -EINVAL;
+}
+
+static void tcp_send_queued_data(struct tcp *conn)
+{
+	struct net_pkt *pkt;
+	int len;
+	int ret;
+
+	if (!conn->pending_data) {
+		/* No pending data in data send queue yet */
+		return;
+	}
+
+	len = net_pkt_get_len(conn->pending_data) - conn->pending_len;
+	if (len < 0) {
+		NET_DBG("Packet %p len %zd is shorter than pending data %zd",
+			conn->pending_data,
+			net_pkt_get_len(conn->pending_data),
+			conn->pending_len);
+		return;
+	}
+
+	/* TODO: We should adjust the amount of data to sent according to
+	 * our number of available buffers etc.
+	 */
+	len = MIN(conn->send_win, len);
+	if (len == 0) {
+		/* All pending data has been sent */
+		NET_DBG("All pending data is sent (conn %p)", conn);
+		return;
+	}
+
+	NET_DBG("Sending %d win %zd pending %zd seq %u ack %u total %zd", len,
+		conn->send_win, conn->pending_len,
+		conn->seq, conn->ack, net_pkt_get_len(conn->pending_data));
+
+	pkt = tcp_pkt_alloc(conn, len);
+	if (!pkt) {
+		NET_DBG("Cannot allocate pkt");
+		return;
+	}
+
+	ret = ip_header_add(conn, pkt);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	ret = tcp_header_add(conn, pkt, PSH | ACK);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	ret = net_pkt_skip(conn->pending_data, conn->pending_len);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	ret = net_pkt_copy(pkt, conn->pending_data, len);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	/* If cursor is reset by net_pkt_copy(), it means that we copied all
+	 * data from it.
+	 */
+	if (!conn->pending_data->cursor.buf) {
+		NET_DBG("Copied all data (%d bytes) from net_pkt %p (len %zd)",
+			len, conn->pending_data,
+			net_pkt_get_len(conn->pending_data));
+		net_pkt_cursor_init(conn->pending_data);
+	}
+
+	ret = tcp_finalize_pkt(pkt);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	conn->pending_len += len;
+	conn_seq(conn, + len);
+
+	NET_DBG("%s", log_strdup(tcp_th(pkt)));
+
+	if (!k_delayed_work_remaining_get(&conn->send_timer)) {
+		k_delayed_work_submit(&conn->send_timer, K_MSEC(tcp_rto));
+	}
+
+	tcp_send(pkt);
+	return;
+
+fail:
+	tcp_pkt_unref(pkt);
+}
+
 static void tcp_send_process(struct k_work *work)
 {
 	struct tcp *conn = CONTAINER_OF(work, struct tcp, send_timer);
 	struct net_pkt *pkt = tcp_slist(&conn->send_queue, peek_head,
 					struct net_pkt, next);
+	if (!pkt) {
```

Comment:
tcp_send_process() works as following:
1. packet is added to the tail of the send_queue by tcp_out()
2. tcp_send_process() is invoked and will start the retransmission
3. tcp_send_timer_cancel() cancels the retransmission for the packet at the head of queue and removes it from the send_queue
4. If there are more packets in the queue, the next packet at the head will be retrasmitted.
5. Packets which are just ACK, PSH, ACK+PSH, RST are not retransmitted.

Working with the send_timer outside of tcp_send_process(), tcp_send_timer_cancel() will interfere with the logic for sending non-data through it.

To resolve this, either:

1. Work with it through tcp_out()
2. Add and use another timer
3. Re-work tcp_send_process() logic of handling data